### ISSUE TITLE
Using NoisyOr as CPDs of BayesianNetwork

### ISSUE THREAD

**Issue by nickpuntoforhof** [pgmpy/pgmpy#1386]
### Subject of the issue
The TabularCPD class is baked into the BayesianEstimator class.  For our research, we are looking to use Noisy-Or models as the conditional probability distributions for each node in a bayesian network.  While a Noisy-Or model class exists in pgmpy, it cannot be used as the CPD for a network model, along with all the functionality this requires.  

I have forked the pgmpy repository and plan on adding this functionality myself.

What I plan on adding

- Write NoisyOr class from scratch within pgmpy.factors.discrete.CPD
        - Replicate TabularCPD class api to prevent bugs.
        - Subclass of DiscreteFactor
        - Use existing code in pgmpy.models.NoisyOrModel to derive functionality of this class.
- Changes to BayesianEstimator, MaximumLikelihoodEstimator
        - Add functionaltiy to .getParameters method
        - Add argument for NoisyOr and ifelse statement to pass node, edge, data to NoisyOr class
        - Add NoisyOr learning functionality to both estimators using pgmpy.estimators.base methods to get sufficient statistic
- Changes to BayesianModel
        - Allow a NoisyOr class to be passed to add_cpds() without error
        - Update comments/documentation

**Comment by ankurankan** [comment]
@nickpuntoforhof Thanks for working on adding this feature. I have a few suggestions for the implementation:

- I would suggest keeping the model structure and parameterization code separate as it helps in keep things modular. So, instead of creating a `NoisyOr` class in `pgmpy.factors.discrete.CPD`, I would suggest creating a simpler `NoisyOrCPD` class that inherits the `TabularCPD` class and adds the restrictions of a NoisyOr CPD. 
- Since the BayesianModel class can represent a NoisyOr model structure as well, I would suggest simply using it without any change. Will make life easier for me in the future as I will have to maintain less code.
- Instead of changing the Estimator methods, I would suggest adding `get_parameters` directly in the `NoisyOrCPD` class. I want to do a similar change for `TabularCPD` class such that estimation methods are in the CPD and the estimator classes just call these methods depending on the type of CPD it wants to learn.

So, in short, I would suggest just adding a `NoisyOrCPD` class in `pgmpy.factors.discrete` which implements the restrictions of a NoisyOr model and methods for learning it from the data. Please let me know what you think about this.

**Comment by ThomasHoppe** [comment]
@ankurankan You wrote "Since the BayesianModel class can represent a NoisyOr model structure as well" I ask myself how this is done? 

I usually use BayesianNetworks, but now I need some NoisyOr-Nodes in the network. How do I declare variables in my network, as noisy-or nodes? Unfortunately the examples in Noisy-Or-Model do not help, nor can I find any example of how to represent these nodes in a BayesianNetwork

**Comment by kgoyal40** [comment]
Hey All, I am wondering if the NoisyOrCPD was ever added in pgmpy.factors.discrete?

**Comment by ankurankan** [comment]
pgmpy.factors.discrete.NoisyORCPD has been added.
